Add feature to ban the participants from participating in a challenge

Sanji515 opened the pull request:
#### Changes proposed in this pull request:

- Added one more new tab named `Settings` on the challenge page for the host user

- Listed all the banned email ids with pencil edit icon. On clicking it, a text box with the all the banned email ids as it's value will be shown.

- When editing there will be two options (icons), one to save the changes and another one (cross icon) to discard the changes

- Link to live demo: http://pr-163-evalai.surge.sh

**<p align="center">Screenshot 1:</p>**


![Screenshot from 2019-07-28 21-02-58](https://user-images.githubusercontent.com/32524438/62009299-55cb6e80-b17b-11e9-9c39-994d4814e1b7.png)



**<p align="center">Screenshot 2:</p>**

![Screenshot from 2019-07-28 21-03-48](https://user-images.githubusercontent.com/32524438/62009301-5c59e600-b17b-11e9-8b91-b5182e72624f.png)

Sanji515 committed:
Add feature to ban the participants from participating in a challenge

Added one more new tab names settings on the challenge page for the host user
Added a text box and which will be shown on click of edit pencil icon
When editing there will be two icons, one to save the changes and another cross icon to discard the changes

RishabhJain2018 commented:
@Sanji515 Please resolve the conflicts here.

lunayach commented:
@Sanji515 resolve the conflicts here.

Sanji515 committed:
resolved conflicts

Sanji515 committed:
Added default None messge when there are no banned email ids

Sanji515 committed:
resolved the tests failing errors

Sanji515 committed:
Merge branch 'master' into ban-participant

Sanji515 commented:
@lunayach @galipremsagar @Shekharrajak Please review

galipremsagar commented:
@Sanji515 resolve conflicts?

Sanji515 committed:
resolved conflicts

Sanji515 committed:
Merge branch 'ban-participant' of https://github.com/Sanji515/EvalAI-ngx into ban-participant

Sanji515 commented:
@galipremsagar I've resolved the conflicts, please review.

galipremsagar committed:
Merge branch 'master' into ban-participant

galipremsagar commented on src/app/components/challenge/challengesettings/challengesettings.component.html:
```diff
@@ -0,0 +1,55 @@
+      </div>
+    </div>
+  </div>
+</div>
```
Add a new-line at the end of all newly added files.
```suggestion
</div>

```

Sanji515 replied:
@galipremsagar done

Sanji515 committed:
added new line

Sanji515 committed:
Merge branch 'ban-participant' of https://github.com/Sanji515/EvalAI-ngx into ban-participant

Sanji515 committed:
resolved conflicts

galipremsagar committed:
Merge branch 'master' into ban-participant

galipremsagar commented on src/app/components/challenge/challengesettings/challengesettings.component.ts:
```diff
@@ -0,0 +1,81 @@
+  getBannedEmailsListObject(bannedEmailInputValue) {
+    const bannedEmailIds = [];
+    // First remove all the whitespaces and then split
+    const splittedEmails = bannedEmailInputValue.replace(/\s/g, '').split(',');
```
`replace(/\s/g, '')`
Why this replace?

Sanji515 replied:
@galipremsagar to remove all the whitespaces like suppose the list is like `abc@gmail.com, xyz@gmail.com` so I don't want the space after comma `,`.

galipremsagar replied:
we can do this: 

```suggestion
    const splittedEmails = bannedEmailInputValue.split(",").map(item => item.trim());
```

Sanji515 replied:
done

galipremsagar reviewed (changes requested):
If the input is empty and we click on green tick it doesn't accept. Either we should we showing some error or just accept the input and process it.

Sanji515 committed:
Merge branch 'master' into ban-participant

Sanji515 committed:
suggested change

lunayach commented on src/app/components/challenge/challengesettings/challengesettings.component.html:
```diff
@@ -0,0 +1,55 @@
+
```
Remove this empty line.

Sanji515 replied:
@lunayach done

lunayach commented on src/app/components/challenge/challengesettings/challengesettings.component.ts:
```diff
@@ -0,0 +1,81 @@
+          SELF.globalService.handleApiError(err, true);
+          SELF.globalService.showToast('error', err);
+        },
+        () => console.log('BANNED-EMAIL-IDS-UPDATE-FINISHED')
```
Remove this.

Sanji515 replied:
@lunayach done

Sanji515 committed:
removed empty line and console statement

lunayach commented on src/app/components/challenge/challengesettings/challengesettings.component.html:
```diff
@@ -0,0 +1,54 @@
+          <span *ngIf="!isBannedEmailInputVisible" class="display-inline">
+            <span class="row">
+              <span *ngIf="challenge.banned_email_ids" class="col-md-7 col-sm-7 li-mr">
+                <span style="white-space: pre-line;" *ngFor="let participant_email of challenge.banned_email_ids"
```
Add `style` to `.scss` file.

Sanji515 replied:
@lunayach Removed it as there's no need of it as now used `chips` to show the banned emails, thanks :)

lunayach commented on src/app/components/challenge/challengesettings/challengesettings.component.ts:
```diff
@@ -0,0 +1,81 @@
+    // First remove all the whitespaces and then split
+    const splittedEmails = bannedEmailInputValue.split(',').map(item => item.trim());
+    for (let i = 0; i < splittedEmails.length; i++) {
+      bannedEmailIds.push(splittedEmails[i]);
```
Should we also apply an email `regex` check for each email?

Sanji515 replied:
@lunayach done

lunayach commented:
As we add email ID's one by one, should we have something like this (https://stackblitz.com/angular/earboyoxrglg?file=app%2Fchips-input-example.ts). That is, each email is added as a chip. (IDK, but this looks more intuitive)

Sanji515 committed:
added chips for each banned email ids

Sanji515 committed:
added css for the input

Sanji515 committed:
Merge branch 'master' into ban-participant

Sanji515 commented:
> As we add email ID's one by one, should we have something like this (https://stackblitz.com/angular/earboyoxrglg?file=app%2Fchips-input-example.ts). That is, each email is added as a chip. (IDK, but this looks more intuitive)

Yes, it looks great. Implemented it and removed the `tick` and `cross` icon. Please take a look @lunayach @RishabhJain2018

Sanji515 committed:
removed unused css

Sanji515 committed:
Merge branch 'ban-participant' of https://github.com/Sanji515/EvalAI-ngx into ban-participant

galipremsagar reviewed (approved):
Changes LGTM, This can be merged.

galipremsagar commented:
@Sanji515  Just update the screenshots of the UI you have posted above as it looks much nicer and better now

RishabhJain2018 committed:
Merge branch 'master' into ban-participant

Sanji515 committed:
uUdated challenge settinngs specs
